Client/Frontend/Browser/TabScrollController.swift

```diff
@@ -198,18 +199,35 @@ private extension TabScrollingController {
         return y
     }
 
-    func animateToolbarsWithOffsets(animated animated: Bool, duration: NSTimeInterval, headerOffset: CGFloat,
+    func animateToolbarsWithOffsets(animated animated: Bool, duration: NSTimeInterval, headerOffset: CGFloat, scrollOffset: CGFloat,
         footerOffset: CGFloat, alpha: CGFloat, completion: ((finished: Bool) -> Void)?) {
 
+        // Due to the header not being placed on top of the scroll view, when it animates away, and the scroll offset is negative,
+        // its superview background will show through temporarily, leading to anpleasant visual effect. By inserted a subview of the
+        // same colour as the background, we can hide this effect for the duration of the animation.
+        var veil: UIView?
+        if let header = self.header, superview = header.superview where animated {
```

Comment:
> Yeah, this was a gnarly issue. 

Yup I feel your pain. When I initially worked on getting the URL/toolbars to scroll it was a nightmare - it's not an easy task 😖

> The problem is that previously, the web content was linked by a constraint to the URL bar, and so when the URL bar animated away, the web content would maintain its relative offset to the URL bar, meaning the bar didn't lift off the web content, and the black bar did not appear.
Now, the URL bar is independent of the content, with regards to its positioning.

Got it. Figured that was what was happening.

> A timing based approach would be difficult to pull off. For the graphical issue to be resolved with that approach, we would want to move the URL bar only when the web content starts pushing it. This would not be immediate, and would be variable depending on the amount the user has scrolled up above the page — this means we would need to start the URL bar animation at a different time to the main animation.

I wonder if we can utilize [this](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIScrollViewDelegate_Protocol/#//apple_ref/occ/intfm/UIScrollViewDelegate/scrollViewWillEndDragging:withVelocity:targetContentOffset:) delegate method to determine this. If we know where the user will end up and the velocity, we might be able to factor those into when we show the URL bar. Still doesn't solve the problem of scrolling the web view at time T1 and starting the animation at a later time T2 where you don't really know what the delta is between those two times. 

>  What's more, the animation is not linear, so in order for it to line up precisely, we would need to start the URL bar animation at a different time, with a different timing function that depended on how far the user had scrolled up (or, equivalent, the velocity of the bounce).

I think the non-linear nature of this makes it more difficult and was something I didn't factor in when considering the timing function. 

All in all, maybe check to see if you can use that delegate for something (if you haven't already) and if not I'm good with landing the veil approach. Thanks for taking a hard look at this - as I mentioned before it's not easy!